avocado/utils/asset.py

```diff
@@ -199,127 +299,105 @@ def fetch(self):
             except Exception:  # pylint: disable=W0703
                 exc_type, exc_value = sys.exc_info()[:2]
-                log.error('%s: %s', exc_type.__name__, exc_value)
+                LOG.error('%s: %s', exc_type.__name__, exc_value)
 
-        raise EnvironmentError("Failed to fetch %s." % basename)
+        # If we get here, we could not download the asset.
+        raise EnvironmentError("Failed to fetch %s." % self.basename)
 
-    def _download(self, url_obj, asset_file):
-        try:
-            # Temporary unique name to use while downloading
-            temp = '%s.%s' % (asset_file,
-                              next(tempfile._get_candidate_names()))  # pylint: disable=W0212
-            url_download(url_obj.geturl(), temp)
-
-            # Acquire lock only after download the file
-            with FileLock(asset_file, 1):
-                shutil.copy(temp, asset_file)
-                self._compute_hash(asset_file)
-                return self._verify(asset_file)
-        finally:
-            os.remove(temp)
-
-    def _compute_hash(self, asset_file):
-        result = crypto.hash_file(asset_file, algorithm=self.algorithm)
-        with open(self._get_hash_file(asset_file), 'w') as f:
-            f.write('%s %s\n' % (self.algorithm, result))
-
-    def _get_hash_from_file(self, asset_file):
-        discovered = None
-        hash_file = self._get_hash_file(asset_file)
-        if not os.path.isfile(hash_file):
-            self._compute_hash(asset_file)
-
-        with open(hash_file, 'r') as hash_file:
-            for line in hash_file:
-                # md5 is 32 chars big and sha512 is 128 chars big.
-                # others supported algorithms are between those.
-                pattern = '%s [a-f0-9]{32,128}' % self.algorithm
-                if re.match(pattern, line):
-                    discovered = line.split()[1]
-                    break
-        return discovered
+    def find_asset_file(self):
+        """
+        Search for the asset file in each one of the cache locations
 
-    def _verify(self, asset_file):
-        if not self.asset_hash:
-            return True
-        if self._get_hash_from_file(asset_file) == self.asset_hash:
-            return True
-        else:
-            return False
+        :return: asset file if exists or None
+        :rtype: str or None
+        """
+        for cache_dir in self.cache_dirs:
+            cache_dir = os.path.expanduser(cache_dir)
 
-    def _get_local_file(self, url_obj, asset_file):
-        if os.path.isdir(url_obj.path):
-            path = os.path.join(url_obj.path, self.name)
-        else:
-            path = url_obj.path
+            for location in self.locations:
+                relative_dir = self._get_relative_dir(location)
+                asset_path = os.path.join(cache_dir,
+                                          relative_dir,
+                                          self.basename)
```

Comment:
In `fetch()` you have this same pattern: loop for locations, getting the asset_path. Perhaps you can convert it in a generator? Not a big deal though...